src/client/core/prevent-real-events.js

```diff
@@ -0,0 +1,71 @@
+import { utils, eventSandbox } from './deps/hammerhead';
+
+import { get, hasDimensions } from './utils/style';
+import { filter } from './utils/array';
+import { isShadowUIElement, isWindow, getParents } from './utils/dom';
+
+var browserUtils   = utils.browser;
+var listeners      = eventSandbox.listeners;
+var eventSimulator = eventSandbox.eventSimulator;
+
+const PREVENTED_EVENTS = [
+    'click', 'mousedown', 'mouseup', 'dblclick', 'contextmenu', 'mousemove', 'mouseover', 'mouseout',
+    'touchstart', 'touchmove', 'touchend', 'keydown', 'keypress', 'input', 'keyup', 'change', 'focus', 'blur',
+    'MSPointerDown', 'MSPointerMove', 'MSPointerOver', 'MSPointerOut', 'MSPointerUp', 'pointerdown',
+    'pointermove', 'pointerover', 'pointerout', 'pointerup'
+];
+
+const F12_KEY_CODE = 123;
+
+
+function checkDevToolsShortcut (e) {
+    // NOTE: Browser tools opening by F12, CTRL+SHIFT+<SYMBOL KEY> on PC or by OPTION(ALT)+CMD+<SYMBOL KEY> on Mac.
+    return e.shiftKey && e.ctrlKey || e.altKey && e.isMacPlatform || e.keyCode === F12_KEY_CODE;
+}
+
+// NOTE: when test is run we should block real events (from mouse, keyboard), because it may lead to
+// unexpected test result.
+function preventRealEventHandler (e, dispatched, preventDefault, cancelHandlers, stopEventPropagation) {
+    var target = e.target || e.srcElement;
+
+    // NOTE: this will allow to press hotkeys to open developer tools
+    if (/^key/.test(e.type) && checkDevToolsShortcut(e)) {
+        stopEventPropagation();
+        return;
+    }
+
+    if (!dispatched && !isShadowUIElement(target)) {
+        //If an element loses focus because of becoming invisible, a blur event is raised. We must not prevent this blur event.
+        //In the IE an element loses focus only if css display property is set to 'none', other ways of making
+        // element invisible don't lead to blurring (in MSEdge focus/blur is sync)

```

Comment:
```
        // If an element loses focus because of becoming invisible, a blur event is raised. We must not prevent this blur event.
        // In IE, an element loses focus only if the CSS 'display' property is set to 'none', other ways of making
        // an element invisible don't lead to blurring (in MSEdge, focus/blur are sync)
```